packages/app/pages/course/[cid]/insights.tsx

```diff
@@ -151,40 +161,97 @@ interface RenderInsightProps {
   insightName: string;
   insightDisplay: InsightDisplayInfo;
   dateRange: DateRangeType;
+  mostActiveStudentsPage: number;
+  setMostActiveStudentsPage: Dispatch<SetStateAction<number>>;
 }
 
+const equalRenderInsights = (
+  prevProps: RenderInsightProps,
+  nextProps: RenderInsightProps
+): boolean => {
+  if (
+    prevProps.insightName === "MostActiveStudents" &&
+    nextProps.insightName === "MostActiveStudents"
+  ) {
+    return (
+      prevProps.mostActiveStudentsPage === nextProps.mostActiveStudentsPage &&
+      prevProps.dateRange.start === nextProps.dateRange.start &&
+      prevProps.dateRange.end === nextProps.dateRange.end
+    );
+  } else {
+    return (
+      prevProps.insightName === nextProps.insightName &&
+      prevProps.dateRange.start === nextProps.dateRange.start &&
+      prevProps.dateRange.end === nextProps.dateRange.end
+    );
+  }
+};
+const MemoizedRenderInsight = React.memo(RenderInsight, equalRenderInsights);
```

Comment:
i dont know much about the memoization in react but this seems pretty slick!

Reply:
Yeah react having that in the library was clutch for a couple reasons
1. Every time you paginate the table without the memoize, it has to rerender the component, which means it has to rerender the parent… which in turn rerenders every insight lol. That’s pretty inefficient when only one component is changing, this allows the old insights to be reused for every rerender where they don’t change.
2. On top of that, the insight component makes the API request to the backend every time it’s rerendered, so memoizing it saves a lot of traffic

Reply:
that is very convenient, nice!